rfd/0141-tsh-bench-db.md

```diff
@@ -0,0 +1,339 @@
+| `database`  | Teleport target database name. Available databases can be retrieved by running `tsh db ls`. If not provided, the `--uri` flag (for direct database connection) must be present. | 
+
+Note: The `--rate` common flag will indicate how many connections per second will be
+established.
+
+**Example:**
+```code
+$ tsh bench db connect --db-user=postgres --db-name=postgres postgres-dev
+```
+
+```mermaid
+sequenceDiagram
+  participant tsh
+  participant Teleport
+  participant DB as Database
+
+  tsh->>Teleport: Generate Certificates (tsh db login)
+
+  loop every execution
+    tsh->>Teleport: Establish connection
+    Teleport->>DB: Connect
+    DB->>tsh: Connection
+    tsh->>DB: Ping
+  end
+```
+
+#### Query
+This suite will test and measure query executions only. Internally it will use a
+connection pool that will be initialized before it starts, meaning that the
+connection time will not be counted on the benchmark results. Users can provide
+setup instructions to create the necessary database structure and load data,
+which won't be counted towards the final results.
+
+Having a flexible command like this enables developers to experiment with
+different scenarios besides relying on adding benchmark subcommands. For
+example, the same command can be used to measure multiple large amounts of
+connections constantly sending queries that return data (such as `SELECT 1;),
+but also can have a very complex query that returns a large chunk of data.
+
+After a general use, we can asses if more subcommands should be added to cover
+the most common use cases, making running benchmarks more convenient and less
+error-prone.
+
+```code
+tsh bench db query [--schema] [--db-user] [--uri=] [--connections=] [--create-sql=] [--query-sql=] [database]
+```
+
+| Flag            | Description |
+| --------------- | ----------- |
+| `--schema`      | If not provided the queries will be executed on the database's default schema. |
+| `--db-user`     | Database user used to connect to the target database. The user must have enough permissions on the database to execute all the benchmark queries. |
+| `--uri`         | Direct database access URI. When provided, the benchmark will issue connections directly to this database, and no Teleport is involved in the testing. It must contain all the connection information, including authentication credentials. The contents from --db-user will be ignored when provided. |
+| `--connections` | The number of connections used to run the benchmark queries. If not provided, the value will be the same as `--rate`. |
+| `--create-sql`  | Specifies literal contents or a path to a SQL file to create database tables and load them with data. |
+| `--query-sql`   | Specifies literal contents or a path to a SQL file that contains query statements to benchmark against the database. |
```

Comment:
Yea agree. @gabrielcorado  lets remove this from initial scope and keep the first version very simple as we discussed last time and measure only overhead added by teleport  by doing simple "ping" test.

Reply:
Agree the initial version doesn't need to cover it. But I still see good value in having it. To give a short example, it would enable us to benchmark how Teleport behaves when large queries are sent. I've been looking at a SQL server performance issue that only happens when the client sends large queries. I couldn't determine where the bottleneck is, but removing some parts of the process (e.g., skipping audit logging those queries) showed considerable improvement. In the end, I've written a short Golang program to execute those queries to measure the end-user effects.